macro-pdfviewer-api/src/main/java/com/xwiki/pdfviewer/internal/filter/MJSMimeTypeRegistrar.java

```diff
@@ -0,0 +1,175 @@
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.pdfviewer.internal.filter;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarOutputStream;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.apache.tika.Tika;
+import org.apache.tika.config.TikaConfig;
+import org.apache.tika.mime.MediaType;
+import org.apache.tika.mime.MimeTypes;
+import org.apache.tika.mime.MimeTypesReader;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.tika.internal.TikaUtils;
+
+/**
+ * If the correct mime type for files with the .mjs extension is missing, register the appropriate mime type.
+ * To be removed after used platform Tika version includes this fix:
+ * https://github.com/apache/tika/commit/ae737cd2625b5e2659c20c27713785df8bfc1957
+ *
+ * @version $Id$
+ * @since 2.6
+ */
+@Component(roles = MJSMimeTypeRegistrar.class)
+@Singleton
+public class MJSMimeTypeRegistrar
+{
+    private static final String EXCLAMATION_MARK = "!";
+
+    private static final String CUSTOM_TYPE = "custom-mimetypes.xml";
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * If the correct mime type for .mjs files is not detected, uses reflection to modify the {@link TikaUtils}
+     * {@link Tika} object with the right mime type configuration for .mjs files. Also attempts to automatically
+     * modify the appropriate jar file and add the custom-mimetypes.xml file, so that after a future restart,
+     * reflection is no longer needed.
+     */
+    public void maybeRegisterMJSMimeType()
+    {
+        try {
+            String mjsMimeType = TikaUtils.detect("test.mjs");
+            if (!mjsMimeType.equalsIgnoreCase("application/javascript") && !mjsMimeType.equalsIgnoreCase(
+                "text/javascript"))
+            {
+                boolean uploadedCustomMimeType = uploadCustomMimeType();
```

Comment:
I'm not fond of modifying JAR files at runtime. Even if you do this the server still needs to be restarted because you can't force reload the class path of the servlet engine. And if the server needs a restart then I find it more safe and clean to ask the administrator to put the ``custom-mimetypes.xml`` file in ``WEB-INF/classes/org/apache/tika/mime`` (in the XWik WAR).